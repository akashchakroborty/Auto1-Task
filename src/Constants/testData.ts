import { CarsResponse } from '../redux/types/carsTypes';

export const MOCK_CARS_DATA: CarsResponse = {
  cars: [
    {
      stockNumber: 97074,
      manufacturerName: 'Porsche',
      modelName: '718 Cayman',
      color: 'black',
      mileage: { number: 100056, unit: 'km' },
      fuelType: 'Diesel',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 12876,
      manufacturerName: 'Tesla',
      modelName: 'Roadster',
      color: 'yellow',
      mileage: { number: 100481, unit: 'km' },
      fuelType: 'Petrol',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 18029,
      manufacturerName: 'Fiat',
      modelName: '131',
      color: 'white',
      mileage: { number: 100710, unit: 'km' },
      fuelType: 'Petrol',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 65672,
      manufacturerName: 'Skoda',
      modelName: 'Felicia',
      color: 'black',
      mileage: { number: 100751, unit: 'km' },
      fuelType: 'Petrol',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 42578,
      manufacturerName: 'Mercedes-Benz',
      modelName: 'Sprinter',
      color: 'yellow',
      mileage: { number: 100782, unit: 'km' },
      fuelType: 'Diesel',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 70917,
      manufacturerName: 'Tesla',
      modelName: 'Roadster',
      color: 'green',
      mileage: { number: 100929, unit: 'km' },
      fuelType: 'Diesel',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 53029,
      manufacturerName: 'Tesla',
      modelName: 'Model X',
      color: 'black',
      mileage: { number: 100997, unit: 'km' },
      fuelType: 'Diesel',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 77489,
      manufacturerName: 'Chrysler',
      modelName: 'Viper',
      color: 'red',
      mileage: { number: 101001, unit: 'km' },
      fuelType: 'Petrol',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 98739,
      manufacturerName: 'Tesla',
      modelName: 'Roadster',
      color: 'blue',
      mileage: { number: 101066, unit: 'km' },
      fuelType: 'Petrol',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 72351,
      manufacturerName: 'Porsche',
      modelName: '968',
      color: 'green',
      mileage: { number: 101069, unit: 'km' },
      fuelType: 'Diesel',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
  ],
  totalPageCount: 100,
  totalCarsCount: 1000,
};

export const MOCK_COLORS = {
  colors: ['red', 'blue', 'green', 'black', 'yellow', 'white', 'silver'],
};

export const MOCK_MANUFACTURERS = {
  manufacturers: [
    {
      name: 'Audi',
      models: [
        {
          name: '100',
        },
      ],
    },
  ],
};
