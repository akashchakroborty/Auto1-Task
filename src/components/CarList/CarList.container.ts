import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { getCars } from '../../redux/Cars/actionCreators';
import {
  selectCars,
  selectIsCarsLoading,
  selectIsError,
  selectTotalCarsCount,
  selectTotalPageCount,
} from '../../redux/Cars/selector';
import { AppState } from '../../Redux/store';
import { CarsActionTypes } from '../../redux/types/carsActionTypes';
import { CarsRequest } from '../../redux/types/carsTypes';
import CarList from '.';

const mapStateToProps = (state: AppState) => ({
  cars: selectCars(state),
  loading: selectIsCarsLoading(state),
  error: selectIsError(state),
  totalCarsCount: selectTotalCarsCount(state),
  totalPageCount: selectTotalPageCount(state),
});

const mapDispatchToProps = (dispatch: ThunkDispatch<AppState, void, CarsActionTypes>) => {
  return {
    getCars: (props: CarsRequest) => {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      dispatch(getCars(props));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(CarList);
