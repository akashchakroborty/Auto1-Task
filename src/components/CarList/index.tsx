import React, { useEffect } from 'react';
import { Typography, Alert } from '@mui/material';
import { Car, CarsRequest } from '../../redux/types/carsTypes';

export interface CarListProps {
  getCars: (props: CarsRequest) => void;
  cars: Car[];
  error: boolean;
  totalCarsCount: number;
}

const CarList: React.FC<CarListProps> = ({
 getCars, cars, error, totalCarsCount
}) => {
  useEffect(() => {
    if(!cars.length) {
      getCars({ sort: 'asc', page: 1 });
    }
  }, [getCars]);

  return (
    <>
      <Typography fontWeight={700} variant="h6" component="h1">
        Available cars
      </Typography>
      <Typography variant="h6" component="h1">
        {`Showing ${cars.length} of ${totalCarsCount} results`}
      </Typography>
      {error && <Alert severity="error">An error has occured.</Alert>}
    </>
  );
};

export default CarList;
