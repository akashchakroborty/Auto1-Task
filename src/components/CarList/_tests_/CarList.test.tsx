import React from 'react';
import { shallow } from 'enzyme';
import CarList, { CarListProps } from '..';
import { MOCK_CARS_DATA } from '../../../Constants/testData';

describe('should test the CarList Component', () => {
  const mockProps: CarListProps = {
    setFilters: jest.fn(),
    getCars: jest.fn(),
    cars: MOCK_CARS_DATA.cars,
    error: false,
    totalCarsCount: MOCK_CARS_DATA.totalCarsCount,
    loading: false,
    totalPageCount: MOCK_CARS_DATA.totalPageCount,
    activeFilters: { page: 1, sort: 'asc' },
  };
  beforeEach(() => {
    jest.spyOn(React, 'useEffect').mockImplementation((f) => f());
  });
  test('should render the component', () => {
    const wrapper = shallow(<CarList {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#errorAlert').exists()).toBe(false);
    expect(wrapper.find('CarListItemSkeleton').exists()).toBe(false);
  });
  test('should test handlePaginationChange function', () => {
    const wrapper = shallow(<CarList {...mockProps} />);
    wrapper.find('#pagination').simulate('change', 2);
    expect(mockProps.setFilters).toHaveBeenCalled();
    expect(mockProps.getCars).toHaveBeenCalled();
  });
  test('should render the skeleton component in case of loading', () => {
    const newMockProps = { ...mockProps, loading: true };
    const wrapper = shallow(<CarList {...newMockProps} />);
    expect(wrapper.find('CarListItemSkeleton').exists()).toBe(true);
  });
  test('should render the alert component in case of error', () => {
    const newMockProps = { ...mockProps, error: true };
    const wrapper = shallow(<CarList {...newMockProps} />);
    expect(wrapper.find('#errorAlert').exists()).toBe(true);
  });
});
