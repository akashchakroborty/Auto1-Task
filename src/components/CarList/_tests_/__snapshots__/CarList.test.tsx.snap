// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the CarList Component should render the component 1`] = `
<Fragment>
  <ForwardRef(Typography)
    component="h1"
    fontWeight={700}
    variant="h6"
  >
    Available cars
  </ForwardRef(Typography)>
  <ForwardRef(Typography)
    component="h1"
    variant="h6"
  >
    Showing 10 of 1000 results
  </ForwardRef(Typography)>
  <Component
    carItem={
      Object {
        "color": "black",
        "fuelType": "Diesel",
        "manufacturerName": "Porsche",
        "mileage": Object {
          "number": 100056,
          "unit": "km",
        },
        "modelName": "718 Cayman",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg",
        "stockNumber": 97074,
      }
    }
    key="97074-718 Cayman"
  />
  <Component
    carItem={
      Object {
        "color": "yellow",
        "fuelType": "Petrol",
        "manufacturerName": "Tesla",
        "mileage": Object {
          "number": 100481,
          "unit": "km",
        },
        "modelName": "Roadster",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg",
        "stockNumber": 12876,
      }
    }
    key="12876-Roadster"
  />
  <Component
    carItem={
      Object {
        "color": "white",
        "fuelType": "Petrol",
        "manufacturerName": "Fiat",
        "mileage": Object {
          "number": 100710,
          "unit": "km",
        },
        "modelName": "131",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg",
        "stockNumber": 18029,
      }
    }
    key="18029-131"
  />
  <Component
    carItem={
      Object {
        "color": "black",
        "fuelType": "Petrol",
        "manufacturerName": "Skoda",
        "mileage": Object {
          "number": 100751,
          "unit": "km",
        },
        "modelName": "Felicia",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg",
        "stockNumber": 65672,
      }
    }
    key="65672-Felicia"
  />
  <Component
    carItem={
      Object {
        "color": "yellow",
        "fuelType": "Diesel",
        "manufacturerName": "Mercedes-Benz",
        "mileage": Object {
          "number": 100782,
          "unit": "km",
        },
        "modelName": "Sprinter",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg",
        "stockNumber": 42578,
      }
    }
    key="42578-Sprinter"
  />
  <Component
    carItem={
      Object {
        "color": "green",
        "fuelType": "Diesel",
        "manufacturerName": "Tesla",
        "mileage": Object {
          "number": 100929,
          "unit": "km",
        },
        "modelName": "Roadster",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg",
        "stockNumber": 70917,
      }
    }
    key="70917-Roadster"
  />
  <Component
    carItem={
      Object {
        "color": "black",
        "fuelType": "Diesel",
        "manufacturerName": "Tesla",
        "mileage": Object {
          "number": 100997,
          "unit": "km",
        },
        "modelName": "Model X",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg",
        "stockNumber": 53029,
      }
    }
    key="53029-Model X"
  />
  <Component
    carItem={
      Object {
        "color": "red",
        "fuelType": "Petrol",
        "manufacturerName": "Chrysler",
        "mileage": Object {
          "number": 101001,
          "unit": "km",
        },
        "modelName": "Viper",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg",
        "stockNumber": 77489,
      }
    }
    key="77489-Viper"
  />
  <Component
    carItem={
      Object {
        "color": "blue",
        "fuelType": "Petrol",
        "manufacturerName": "Tesla",
        "mileage": Object {
          "number": 101066,
          "unit": "km",
        },
        "modelName": "Roadster",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg",
        "stockNumber": 98739,
      }
    }
    key="98739-Roadster"
  />
  <Component
    carItem={
      Object {
        "color": "green",
        "fuelType": "Diesel",
        "manufacturerName": "Porsche",
        "mileage": Object {
          "number": 101069,
          "unit": "km",
        },
        "modelName": "968",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg",
        "stockNumber": 72351,
      }
    }
    key="72351-968"
  />
  <ForwardRef(Pagination)
    className="makeStyles-pagination-1"
    count={100}
    id="pagination"
    onChange={[Function]}
    page={1}
    showFirstButton={true}
    showLastButton={true}
  />
</Fragment>
`;
